### PR TITLE
Add Decimal class

### PR DESCRIPTION
**Pull request by jodastephen**
* A representation of a Decimal better than BigDecimal
* Internally it closely matches BigDecimal to simplify conversion
* Single representation (no trailing zero fractions)

### CODE REVIEW

File: modules/collect/src/main/java/com/opengamma/strata/collect/Decimal.java [hunk 1]
```diff
@@ -0,0 +1,710 @@
...
+  }
+
+  // creates from a pre-rounded value
+  private static Decimal ofRounded(BigDecimal value) {
+    BigDecimal adjusted = value.stripTrailingZeros();
+    adjusted = adjusted.scale() < 0 ? adjusted.setScale(0) : adjusted;
+    if (adjusted.precision() > MAX_PRECISION) {
+      throw new IllegalArgumentException("Decimal value must not exceed 18 digits of precision at scale 0: " + value);
+    }
+    // lnogValueExact() used for extra safety, it should never actually throw
```
Comment (ashleyheath) [inline modules/collect/src/main/java/com/opengamma/strata/collect/Decimal.java:213]:
```suggestion
    // longValueExact() used for extra safety, it should never actually throw
```

**Review by ashleyheath — CHANGES_REQUESTED** [review]

**Review by ashleyheath — APPROVED** [review]
LGTM; don't know if @cjkent wants to give it a pass too?

**Comment by cjkent** [comment]
> LGTM; don't know if @cjkent wants to give it a pass too?

I'm looking now

File: modules/collect/src/main/java/com/opengamma/strata/collect/Decimal.java [hunk 2]
```diff
@@ -0,0 +1,710 @@
...
+  private static final long MAX_UNSCALED = 999_999_999_999_999_999L;
+  /** Powers of ten. */
+  private static final long[] POWERS = new long[MAX_SCALE + 1];
+  static {
+    POWERS[0] = 1;
+    for (int i = 1; i < POWERS.length; i++) {
+      POWERS[i] = POWERS[i - 1] * 10;
+    }
+  }
+  /** Math context that truncates. */
```
[the remark this answers is not in the document]
Comment (jodastephen) [reply modules/collect/src/main/java/com/opengamma/strata/collect/Decimal.java:47]: IIUC, the IEEE spec recommends truncation and half-rounding to be considered of equal importance. We don't follow that spec exactly, but it is a guide to the problem space. Here, truncation is an obvious approach to take when converting a string for example, and also matches the behaviour of Java when doing something like `(long) 1.7d`.

File: modules/collect/src/main/java/com/opengamma/strata/collect/Decimal.java [hunk 3]
```diff
@@ -0,0 +1,710 @@
...
+import java.util.Arrays;
+import java.util.function.DoubleUnaryOperator;
+import java.util.function.UnaryOperator;
+
+import org.joda.beans.gen.PropertyDefinition;
+import org.joda.convert.FromString;
+import org.joda.convert.ToString;
+
+/**
+ * A decimal number, similar to {@code BigDecimal}, but optimized for the needs of finance.
```
[the remark this answers is not in the document]
Comment (jodastephen) [reply modules/collect/src/main/java/com/opengamma/strata/collect/Decimal.java:21]: It provides a better, cleaner API for developers. In developing the recent data models, I found it was surprising to have nice classes for `Percentage` and `BasisPoints`, but nothing good for a "normal decimal". This class is the response to that. There are secondary benefits in performance and memory usage. Finally, we will be able to convert this class to a Valhalla value type in the future, but `BigDecimal` will probably never be converted for legacy reasons.

**Review by cjkent — APPROVED** [review]

**Pull request merged**